### PR TITLE
Fix 2FA Expense payment checker

### PR DESCRIPTION
**Pull request by kewitz**

### CODE REVIEW

File: server/controllers/users.js [hunk 1]
```diff
@@ -89,10 +89,13 @@ export const signin = (req, res, next) => {
...
   if (twoFactorAuthenticationEnabled && req.remoteUser.twoFactorAuthToken !== null) {
-    const token = req.remoteUser.jwt({ scope: 'twofactorauth' }, auth.TOKEN_EXPIRATION_SESSION);
+    const token = req.remoteUser.jwt(
+      { scope: 'twofactorauth', sesionId: req.jwtPayload?.sessionId },
+      auth.TOKEN_EXPIRATION_SESSION,
+    );
     res.send({ token });
   } else {
-    const token = req.remoteUser.jwt({}, auth.TOKEN_EXPIRATION_SESSION);
+    const token = req.remoteUser.jwt({ sesionId: req.jwtPayload?.sessionId }, auth.TOKEN_EXPIRATION_SESSION);
```
Comment (znarf) [inline server/controllers/users.js:98]:
```suggestion
    const token = req.remoteUser.jwt({ sessionId: req.jwtPayload?.sessionId }, auth.TOKEN_EXPIRATION_SESSION);
```

**Comment by Betree** [comment]
Related: https://github.com/opencollective/opencollective/issues/3781

**Comment by znarf** [comment]
We use Expense payment a lot with the API these days (1kproject) and we also need a solution that works well with so called "API Keys" (Personal Tokens).

**Comment by kewitz** [comment]
> We use Expense payment a lot with the API these days (1kproject) and we also need a solution that works well with so called "API Keys" (Personal Tokens).

@znarf I think requiring 2FA codes from an API key request is against the purpose of creating the API key in the first place.
I suggest we require the 2FA code to create an API key and ignore the 2FA key for actions when using an API key.

WDYT?

**Comment by znarf** [comment]
> I think requiring 2FA codes from an API key request is against the purpose of creating the API key in the first place.

@kewitz I don't understand how you come to that conclusion. I disagree, I think it's a great security feature.

**Comment by kewitz** [comment]
@znarf people usually create an API key to automate their actions. I don't think it is reasonable to expect someone to automate their actions and require them to always be around in order to type a 2FA code.

**Comment by kewitz** [comment]
@znarf just confirmed this implementation is compatible with the 1K Project script.
Notice that since we're not generating tokens for API Key type of calls, there's also no session attributed to these calls.
Meaning that when you use 2FA code in a single API Key call, you'll still be enabling all calls from the same API Key to pay for expenses.

Do you think this is acceptable?

**Comment by znarf** [comment]
> Do you think this is acceptable?

Sure, but I would make it explicit, and not just rely on `sessionId` being empty. https://github.com/opencollective/opencollective-api/pull/7460/files#diff-7c3df5f093276d2cc8207c916aa13f831d4e6bb7dbfc32bab603a64aefe93b03R34

**Comment by kewitz** [comment]
Rebased.

**Review by znarf — APPROVED** [review]
That should works.

Secondary review from @Betree required.

**Review by Betree — APPROVED** [review]
Looks great and I like that it doesn't introduce too much complexity

**Comment by kewitz** [comment]
@Betree, I was thinking about adding session control in this PR but I'll just create a new issue later on so we can elaborate on it.

**Comment by Betree** [comment]
> @Betree, I was thinking about adding session control in this PR but I'll just create a new issue later on so we can elaborate on it.

@kewitz There's already [an issue](https://github.com/opencollective/opencollective/issues/3781) for that with some raw specs proposal

**Pull request merged**